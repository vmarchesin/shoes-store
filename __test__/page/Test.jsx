
import { Provider, Provider as ProviderRedux, useDispatch } from 'react-redux'
import { createStore, applyMiddleware } from "redux";
import thunkMiddleware from "redux-thunk";

import { render, screen } from "../test-utils"

const initialState = {
  email: "",
  cartlist: [],
  orders: [],
  wishlist: []
}

export const addShoeToWishList = (data) => {
  return dispatch => {
    dispatch({
      type: 'ADD_SHOE_TO_WISHLIST',
      payload: data
    })
  }
}

const reducer = (state, action) => {
  switch (action.type) {
    case 'ADD_SHOE_TO_WISHLIST':
      return {
        ...state,
        wishlist: [...state.wishlist, action.payload]
      }
    default:
      return state
  }
}

const store = createStore(
  reducer,
  initialState,
  applyMiddleware(thunkMiddleware)
);

export default () => {
  const dispatch = useDispatch()

  const wishlist = () => dispatch(addShoeToWishList('shoe'))
  return (
    <ProviderRedux store={store}>
      <div>
        <h1>Title</h1>
        <button onClick={wishlist}>Wishlist</button>
      </div>
    </ProviderRedux>
  )
}