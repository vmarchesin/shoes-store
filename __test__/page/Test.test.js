/** @jest-environment jsdom */
import React from "react";
import Test from './Test'
import { Provider, Provider as ProviderRedux } from 'react-redux'
import renderer from 'react-test-renderer'

import { createStore, applyMiddleware } from "redux";
import thunkMiddleware from "redux-thunk";

import { render, screen } from "../test-utils"

const initialState = {
  email: "",
  cartlist: [],
  orders: [],
  wishlist: []
}

export const addShoeToWishList = (data) => {
  return dispatch => {
    dispatch({
      type: 'ADD_SHOE_TO_WISHLIST',
      payload: data
    })
  }
}

export const resetStore = () => {
  return dispatch => {
    dispatch({
      type: 'RESET_STORE',
    })
  }
}

const reducer = (state, action) => {
  switch (action.type) {
    case 'ADD_SHOE_TO_WISHLIST':
      return {
        ...state,
        wishlist: [...state.wishlist, action.payload]
      }
    case 'RESET_STORE':
      return initialState
    default:
      return state
  }
}

const store = createStore(
  reducer,
  initialState,
  applyMiddleware(thunkMiddleware)
);

beforeEach(() => {
  store.dispatch(resetStore())
})

describe('Test-', () => {
  it('renders a heading', () => {
    render(<ProviderRedux store={store}><Test /></ProviderRedux>)

    const heading = screen.getByText(/Title/i);

    expect(heading).toBeInTheDocument()
  })

  it('adds a shoe to wishlist on store', () => {
    render(<ProviderRedux store={store}><Test /></ProviderRedux>)
    store.dispatch(addShoeToWishList('shoe'))
    expect(store.getState()).toEqual({ email: '', cartlist: [], orders: [], wishlist: [ 'shoe' ] })
  })

  it('adds a shoe to wishlist on click', () => {
    render(<ProviderRedux store={store}><Test /></ProviderRedux>)

    const wishlistButton = screen.getByText(/Wishlist/i);
    expect(wishlistButton).toBeInTheDocument()
    wishlistButton.click()

    expect(store.getState()).toEqual({ email: '', cartlist: [], orders: [], wishlist: [ 'shoe' ] })
  })

  it('matches snapshot', () => {
    const tree = renderer
      .create(<ProviderRedux store={store}><Test /></ProviderRedux>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })
})