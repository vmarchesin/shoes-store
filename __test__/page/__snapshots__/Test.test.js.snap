// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test- matches snapshot 1`] = `
<div>
  <h1>
    Title
  </h1>
  <button
    onClick={[Function]}
  >
    Wishlist
  </button>
</div>
`;
